frame/referenda/src/types.rs

```diff
@@ -307,12 +422,27 @@ impl Curve {
 	/// ```
 	pub fn delay(&self, y: Perbill) -> Perbill {
 		match self {
-			Self::LinearDecreasing { begin, delta } =>
-				if delta.is_zero() {
-					*delta
+			Self::LinearDecreasing { length, floor, ceil } =>
+				if y < *floor {
+					Perbill::one()
+				} else if y > *ceil {
+					Perbill::zero()
 				} else {
-					(*begin - y.min(*begin)).min(*delta) / *delta
+					(*ceil - y).saturating_div(*ceil - *floor, Up) * *length
```

Comment:
Use `mul` or `saturating_mul` here?

```suggestion
					(*ceil - y).saturating_div(*ceil - *floor, Up).mul(*length)
```